lib/ads/interstitial_ad_manager.js

```diff
@@ -0,0 +1,601 @@
+    }
+
+    if (!initialTime) {
+      initialTime = Date.now();
+    }
+
+    if (adPosition == 1 && sequenceLength == 1) {
+      sequenceLength = Array.from(this.interstitials_).filter((i) => {
+        return i.startTime == interstitial.startTime;
+      }).length;
+    }
+
+    this.interstitials_.delete(interstitial);
+    this.playingAd_ = true;
+
+    if (this.usingBaseVideo_ && adPosition == 1) {
+      this.onEvent_(new shaka.util.FakeEvent(
+          shaka.ads.AdManager.AD_CONTENT_PAUSE_REQUESTED,
+          (new Map()).set('saveLivePosition', true)));
+      const detachBasePlayerPromise = new shaka.util.PublicPromise();
+      const checkState = async (e) => {
+        if (e['state'] == 'detach') {
+          if (shaka.util.Platform.isSmartTV()) {
+            await new Promise(
+                (resolve) => new shaka.util.Timer(resolve).tickAfter(0.1));
+          }
+          detachBasePlayerPromise.resolve();
+          this.adEventManager_.unlisten(
+              this.basePlayer_, 'onstatechange', checkState);
+        }
+      };
+      this.adEventManager_.listen(
+          this.basePlayer_, 'onstatechange', checkState);
+      await detachBasePlayerPromise;
+    }
+    if (!this.usingBaseVideo_) {
+      this.baseVideo_.pause();
+    }
+
+    let unloadingInterstitial = false;
+    /** @type {?shaka.util.Timer} */
+    let playoutLimitTimer = null;
+
+    const basicTask = async () => {
+      if (playoutLimitTimer) {
+        playoutLimitTimer.stop();
+      }
+      // Optimization to avoid returning to main content when there is another
+      // interstitial below.
+      const nextCurrentInterstitial = this.getCurrentIntestitial_();
+      if (nextCurrentInterstitial) {
+        this.onEvent_(
+            new shaka.util.FakeEvent(shaka.ads.AdManager.AD_STOPPED));
+        this.adEventManager_.removeAll();
+        this.setupAd_(nextCurrentInterstitial, sequenceLength,
+            ++adPosition, initialTime);
+      } else {
+        if (this.usingBaseVideo_) {
+          await this.player_.detach();
+        } else {
+          await this.player_.unload();
+        }
+        let offset = interstitial.resumeOffset;
+        if (offset == null) {
+          goog.asserts.assert(typeof(initialTime) == 'number',
+              'Should be an number!');
+          offset = (Date.now() - initialTime) / 1000;
+        }
+        if (this.usingBaseVideo_) {
+          this.onEvent_(new shaka.util.FakeEvent(
+              shaka.ads.AdManager.AD_CONTENT_RESUME_REQUESTED,
+              (new Map()).set('offset', offset)));
+        }
+        this.onEvent_(
+            new shaka.util.FakeEvent(shaka.ads.AdManager.AD_STOPPED));
+        this.adEventManager_.removeAll();
+        this.playingAd_ = false;
+        if (!this.usingBaseVideo_) {
+          if (offset) {
+            this.baseVideo_.currentTime += offset;
+          }
+          this.baseVideo_.play();
+        }
+      }
+    };
+    const error = async (e) => {
```

Comment:
Could you make this into a class method, then have the listen to errors on the preload managers using it?
It'd be nice if we could catch an error during the preloading process, and then just not call `player.load` at all, in the case of an ad that's a broken link or whatnot.